### PR TITLE
UDF zero params #5378

### LINKED ISSUE #5378: UDF with zero params broken (doesn't receive null array as input)
**Describe the bug**
For a UDF with zero args, the docs say that the implementation function will receive a single array in its input, consisting of null values, so that the implementation knows the size of the array that needs to be output.

**To Reproduce**
```rust
use std::sync::Arc;

use arrow::{
    array::{ArrayRef, UInt64Array},
    record_batch::RecordBatch,
};
use arrow_schema::DataType;
use datafusion::{
    datasource::MemTable, error::Result, physical_plan::functions::make_scalar_function,
    prelude::SessionContext,
};
use datafusion_expr::{create_udf, Volatility};

#[tokio::main]
async fn main() -> Result<()> {
    let udf_impl = |args: &[ArrayRef]| {
        // according to datafusion docs, this should have a single arg
        // consisting of a null array
        assert_eq!(args.len(), 1); // panics right here
        // use len of input array to know how big the output array needs to be
        let size = args[0].len();
        let output = (0..size).map(|_| 1).collect::<UInt64Array>();
        Ok(Arc::new(output) as ArrayRef)
    };

    let udf = create_udf(
        "no_args",
        vec![],
        Arc::new(DataType::UInt64),
        Volatility::Immutable,
        make_scalar_function(udf_impl),
    );
    let ctx = SessionContext::new();
    ctx.register_udf(udf);
    let batch = RecordBatch::try_from_iter(vec![(
        "col",
        Arc::new((0..5).collect::<UInt64Array>()) as ArrayRef,
    )])?;
    let table = MemTable::try_new(batch.schema(), vec![vec![batch]])?;
    ctx.register_table("test", Arc::new(table))?;
    ctx.sql("select no_args() from test").await?;
    // select no_args() does the same
    Ok(())
}
```
Panics at the assertion in the impl.

**Expected behavior**
What the docs say the behavior will be:
> ...with the exception of zero param function, where a singular element vec
will be passed. In that case the single element is a null array to indicate
the batch's row count (so that the generative zero-argument function can know
the result array size).

**Additional context**
I can look into fixing this one.

### PR DESCRIPTION
**Pull request by jaylmiller**
# Which issue does this PR close?
Closes #5378.

# Rationale for this change

UDFs with zero arguments were not working before.

# What changes are included in this PR?

- Change UDF planning to receive null as input if it takes no args
    - This is following the existing docs specification: 
    - > ...with the exception of zero param function, where a singular element vec
will be passed. In that case the single element is a null array to indicate
the batch's row count (so that the generative zero-argument function can know
the result array size).

- Add test for zero param UDF to sql integration test suite

# Are these changes tested?
Yes.

# Are there any user-facing changes?
Fixing user-facing bug.

### CODE REVIEW

File: datafusion/physical-expr/src/planner.rs [hunk 1]
```diff
@@ -393,7 +393,10 @@ pub fn create_physical_expr(
                     execution_props,
                 )?);
             }
-
+            // udfs with zero params expect null array as input
+            if args.is_empty() {
+                physical_args.push(Arc::new(Literal::new(ScalarValue::Null)));
```
Comment (avantgardnerio) [inline datafusion/physical-expr/src/planner.rs:398]: Though a convenient implementation, I'm wary that generating a fake argument by which to transfer row count will cause problems later.

I'd prefer to see a solution where the row count is passed out-of-band, for example in a change to https://github.com/apache/arrow-datafusion/blob/cef119da9ee8672b1b1e50ac01387dcb1640d96e/datafusion/expr/src/function.rs#L39 that would add an extra argument (i.e. `len: usize`) for this purpose.

If that were present, we could populate it up in the call stack where we know the RecordBatch size, probably here: https://github.com/apache/arrow-datafusion/blob/cef119da9ee8672b1b1e50ac01387dcb1640d96e/datafusion/physical-expr/src/scalar_function.rs#L147

**Review by avantgardnerio — CHANGES_REQUESTED** [review]

File: datafusion/physical-expr/src/planner.rs (hunk 1 above)
Comment (jaylmiller) [reply datafusion/physical-expr/src/planner.rs:398]: Sounds good. Thanks for the suggestions!

Comment (avantgardnerio) [reply datafusion/physical-expr/src/planner.rs:398]: Oh, sorry I missed that. I didn't realize this was part of the existing design.

**Review by avantgardnerio — APPROVED** [review]

File: datafusion/physical-expr/src/planner.rs (hunk 1 above)
Comment (jaylmiller) [reply datafusion/physical-expr/src/planner.rs:398]: No worries I should've put that in the PR itself. Just edited
